### PR TITLE
Make MLflow integration work with kubeflow and scheduled pipelines

### PR DESCRIPTION
**Pull request by stefannica**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
This re-factoring of the MLflow integration is required to make it also function properly in the context of remotely executed pipelines (e.g. kubeflow).

* instead of a pipeline decorator, `enable_mlflow` is now a step decorator, while retaining all of its previous functionality. This is also the only breaking change introduced in this PR. This change was required because the modification made by pipeline decorators to user defined pipelines are not reflected in the compiled pipeline that is executed remotely.
* a StepEnvironment global singleton/context is used to store runtime environment information for the current step that can be consumed from within the step function implementation. The same singleton is used by the MLflow step decorator to determine the current pipeline run ID and pipeline name used in creating the MLflow experiment and run contexts.
* a scheduled pipeline run example is now available for the MLflow integration, in addition to the regular run

### CODE REVIEW

**Comment by htahir1** [comment]
As a general side-question before I even start the review (and feel to just comment on it - i know it is outside of the scope of this particular PR): 

Currently, we are in Scenario 1 of mlflow deployment:
image

Will `mlflow` work with `kubeflow`  in a more production-ready cloud-setting. I.e. When we are in Scenario 4 of mlflow:

image

Source: https://www.mlflow.org/docs/latest/tracking.html#scenario-4-mlflow-with-remote-tracking-server-backend-and-artifact-stores

If any design decisions have been taken in this PR that affect an easy transition into scenario 4, we must ask this question now before merging. Just wanted to bring that up before reviewing (my sincere apologies for that late comment on this issue - i know its a 3rd refactor of this and i should have raised this earlier perhaps)

**Comment by stefannica** [comment]
> If any design decisions have been taken in this PR that affect an easy transition into scenario 4, we must ask this question now before merging. Just wanted to bring that up before reviewing (my sincere apologies for that late comment on this issue - i know its a 3rd refactor of this and i should have raised this earlier perhaps)

@htahir1 This PR is orthogonal with how MLflow can be set up to store artifacts and experiment tracking data. If anything, it makes switching to other MLflow deployment scenarios even more flexible, because now it should theoretically be possible to connect different steps to different MLflow tracking servers and artifact storage backends.

Having a pipeline decorator instead of a step decorator did have one advantage over the current approach: you could enable all steps in a pipeline to MLflow with a single line of code. More of a syntactic convenience, but a useful one nonetheless. Having said that, we can still have both: a step decorator and a pipeline decorator (that basically applies the step decorator to all steps). If you feel we should add that to this PR, let me know, it's not too much effort.

File: examples/mlflow_tracking/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+mlflow>=1.2.0
+tensorflow
```
Comment (AlexejPenner) [inline examples/mlflow_tracking/requirements.txt:2]: Until now we have not used a requirements.txt inside any example besides kubeflow. For kuebflow its teh requirements txt that will be used when building the docker images.

Is there a specific reason you decided to add this here?

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        @enable_mlflow
+        @step
+        def tf_evaluator(
+            x_test: np.ndarray,
+            y_test: np.ndarray,
+            model: tf.keras.Model,
+        ) -> float:
+            _, test_acc = model.evaluate(x_test, y_test, verbose=2)
+            mlflow.log_metric("val_accuracy", test_acc)
+            return test_acc
```
Comment (AlexejPenner) [inline src/zenml/integrations/mlflow/mlflow_step_decorator.py:77]: We are no longer using sphinx api docs - for mkdocs using markdown formatting achieves the same result

```suggestion
    ```python
        @enable_mlflow
        @step
        def tf_evaluator(
            x_test: np.ndarray,
            y_test: np.ndarray,
            model: tf.keras.Model,
        ) -> float:
            _, test_acc = model.evaluate(x_test, y_test, verbose=2)
            mlflow.log_metric("val_accuracy", test_acc)
            return test_acc
    ```
```

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+# step class type
+S = TypeVar("S", bound=Type[BaseStep])
+
+
+@overload
+def enable_mlflow(
+    _step: S,
+) -> S:
+    """Type annotations for mlflow step decorator in case of no arguments."""
```
Comment (AlexejPenner) [inline src/zenml/integrations/mlflow/mlflow_step_decorator.py:39]: For the api docs it might be nice to include an Args: description - not sure if its necessary though

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    ...
+
+
+@overload
+def enable_mlflow(
+    *,
+    experiment_name: Optional[str] = None,
+) -> Callable[[S], S]:
+    """Type annotations for mlflow step decorator in case of arguments."""
```
Comment (AlexejPenner) [inline src/zenml/integrations/mlflow/mlflow_step_decorator.py:48]: Same as above

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+    Returns:
+        The inner decorator which enhaces the input step class with mlflow
+        tracking functionality
+    """
+
+    def inner_decorator(_step: S) -> S:
+
+        logger.debug(
+            f"Applying 'enable_mlflow' decorator to step {_step.__name__}"
```
Comment (AlexejPenner) [inline src/zenml/integrations/mlflow/mlflow_step_decorator.py:99]:
```suggestion
            "Applying 'enable_mlflow' decorator to step %s", _step.__name__"
```

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+            the step is executed will be used as experiment name.
+
+    Returns:
+        the input function enhanced with mlflow profiling functionality
+    """
+
+    def inner_decorator(func: F) -> F:
+
+        logger.debug(
+            f"Applying 'mlflow_step_entrypoint' decorator to step entrypoint {func.__name__}"
```
Comment (AlexejPenner) [inline src/zenml/integrations/mlflow/mlflow_step_decorator.py:140]:
```suggestion
            f"Applying 'mlflow_step_entrypoint' decorator to step entrypoint %s, func.__name__"
```

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        )
+
+        @functools.wraps(func)
+        def wrapper(*args: Any, **kwargs: Any) -> Any:  # noqa
+            logger.debug(
+                f"Setting up MLflow backend before running step entrypoint {func.__name__}"
```
Comment (AlexejPenner) [inline src/zenml/integrations/mlflow/mlflow_step_decorator.py:146]:
```suggestion
                "Setting up MLflow backend before running step entrypoint %s, func.__name__"
```

File: src/zenml/integrations/mlflow/mlflow_utils.py [hunk 8]
```diff
@@ -38,126 +46,59 @@ def setup_mlflow(
...
+    run name.
 
-    In order for a pipeline to run within the context of mlflow, the mlflow
-    experiment should be associated with the pipeline directly. Each separate
-    pipeline run needs to be associated directly with a mlflow experiment. For
-    this, the __init__ and run methods need to be extended accordingly.
+    IMPORTANT: this function is not race condition proof. If two or more
+    processes call it at the same time and with the same arguments, it could
+    lead to a situation where two or more MLflow runs with the same name
+    and different IDs are created.
```
Comment (AlexejPenner) [inline src/zenml/integrations/mlflow/mlflow_utils.py:77]: Might be worth creating a todo or directly creating a ticket to look into this. Feel free to resolve this if its already done

File: src/zenml/steps/step_environment.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        step.
+
+        NOTE: the StepEnvironment class is a singleton. A single
+        StepEnvironment instance exists at any given time and should only
+        be accessed from within a pipeline step function.
+
+        Args:
+            pipeline_name (str): the name of the currently running pipeline
+            pipeline_run_id (str): the ID of the currently running pipeline
+            step_name (str): the name of the currently running step
```
Comment (AlexejPenner) [inline src/zenml/steps/step_environment.py:54]: Our api docs will complain about this notation
```suggestion
            pipeline_name: the name of the currently running pipeline
            pipeline_run_id: the ID of the currently running pipeline
            step_name: the name of the currently running step
```

**Review by AlexejPenner — CHANGES_REQUESTED** [review]
Got some questions and tiny CRs - overall I am super excited about this!

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py (hunk 5 above)
Comment (AlexejPenner) [reply src/zenml/integrations/mlflow/mlflow_step_decorator.py:99]: I know this is probably not really that relevant for us, but I saw you used both approaches in this Pr and thought I'd uniformize it (if thats a word)

File: examples/mlflow_tracking/requirements.txt (hunk 1 above)
Comment (stefannica) [reply examples/mlflow_tracking/requirements.txt:2]: Kubeflow is precisely the reason for this requirements.txt file. I believe we should include one with every example. It makes running examples with the kubeflow orchestrator a lot easier.

**Comment by AlexejPenner** [comment]
> 

Personally, I really liked the @enable_mlflow decorator at the pipeline level and would advocate for implementing it if its not a big hassle.

File: examples/mlflow_tracking/requirements.txt (hunk 1 above)
Comment (AlexejPenner) [reply examples/mlflow_tracking/requirements.txt:2]: Totally makes sense - thanks

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py (hunk 3 above)
Comment (stefannica) [reply src/zenml/integrations/mlflow/mlflow_step_decorator.py:39]: Luckily, the API docs is smart enough to detect and ignore @override cases. We have a similar situation with `enable_whylogs` and the API docs are only rendered for the concrete function: https://apidocs.zenml.io/0.6.0/api_docs/integrations/#zenml.integrations.whylogs.whylogs_step_decorator

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py (hunk 4 above)
Comment (stefannica) [reply src/zenml/integrations/mlflow/mlflow_step_decorator.py:48]: Same as above, not really needed for the API docs. The concrete function already has the full docstrings.

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py (hunk 3 above)
Comment (AlexejPenner) [reply src/zenml/integrations/mlflow/mlflow_step_decorator.py:39]: Nice :)

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py (hunk 2 above)
Comment (stefannica) [reply src/zenml/integrations/mlflow/mlflow_step_decorator.py]: Thanks, nice suggestion !

**Comment by stefannica** [comment]
@AlexejPenner thanks for the CRs, you're a mkdocs / API docs guru. I hope I addressed all your comments.

File: src/zenml/integrations/mlflow/mlflow_step_decorator.py (hunk 2 above)
Comment (AlexejPenner) [reply src/zenml/integrations/mlflow/mlflow_step_decorator.py]: It looks like you changed this for whylogs but net here

**Comment by AlexejPenner** [comment]
> @AlexejPenner thanks for the CRs, you're a mkdocs / API docs guru. I hope I addressed all your comments.

@stefannica You got it all :)

File: src/zenml/environment.py [hunk 10]
```diff
@@ -109,44 +123,262 @@ def in_paperspace_gradient() -> bool:
...
+            assert (
+                cls.NAME and cls.NAME != _BASE_ENVIRONMENT_COMPONENT_NAME
+            ), "You should specify a unique NAME when creating an EnvironmentComponent !"
+        return cls
+
+
+class BaseEnvironmentComponent(metaclass=EnvironmentComponentMeta):
+    """Base Environment component class.
+
+    All Environment components must inherit this class and provide a unique
```
Comment (schustmi) [inline src/zenml/environment.py:260]:
```suggestion
    All Environment components must inherit from this class and provide a unique
```

File: src/zenml/environment.py [hunk 11]
```diff
@@ -109,44 +123,262 @@ def in_paperspace_gradient() -> bool:
...
-        with Environment._layer(...):
-            # only attributes from outer context manager are set
-            with Environment._layer(...):
-                # attributes from outer and inner context manager are set
-                # (inner context manager can overwrite values from outer one)
-
-            # only attributes from outer context manager are set
-        ```
+    def register_component(
+        self, name: str, component: "BaseEnvironmentComponent"
```
Comment (schustmi) [inline src/zenml/environment.py:127]: I guess we could remove the name parameter here and use `component.NAME` instead? Or should we allow registering components with a different name?

File: src/zenml/integrations/mlflow/mlflow_environment.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+        Returns:
+            The MLflow tracking URI for the local mlflow backend.
+        """
+        repo = Repository(root)
+        artifact_store = repo.active_stack.artifact_store
+        local_mlflow_backend_uri = os.path.join(artifact_store.path, "mlruns")
+        if not os.path.exists(local_mlflow_backend_uri):
+            os.makedirs(local_mlflow_backend_uri)
+            # TODO [medium]: safely access (possibly non-existent) artifact stores
```
Comment (schustmi) [inline src/zenml/integrations/mlflow/mlflow_environment.py:81]:
```suggestion
            # TODO [MEDIUM]: safely access (possibly non-existent) artifact stores
```

File: src/zenml/steps/step_environment.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+
+from zenml.environment import BaseEnvironmentComponent
+
+STEP_ENVIRONMENT = "step_environment"
```
Comment (schustmi) [inline src/zenml/steps/step_environment.py:17]: When I first saw the import of that I wasn't quite sure what it was exactly, maybe `STEP_ENVIRONMENT_NAME` to make it a little clearer at first sight?

File: src/zenml/integrations/mlflow/mlflow_environment.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+from mlflow import set_tracking_uri  # type: ignore
+
+from zenml.environment import BaseEnvironmentComponent
+from zenml.logger import get_logger
+from zenml.repository import Repository
+
+logger = get_logger(__name__)
+
+MLFLOW_ENVIRONMENT = "mlflow"
```
Comment (schustmi) [inline src/zenml/integrations/mlflow/mlflow_environment.py:26]: Same as STEP_ENVIRONMENT, maybe add a `_NAME` here?

File: src/zenml/integrations/mlflow/mlflow_utils.py [hunk 15]
```diff
@@ -11,154 +11,57 @@
...
-            (pipeline,),
-            {
-                "__init__": enable_mlflow_init(
-                    pipeline.__init__, experiment_name
-                ),
-                "run": enable_mlflow_run(pipeline.run),
-            },
+    mlflow_experiment = get_experiment_by_name(experiment_name)
+
+    # TODO [medium]: find a solution to avoid race-conditions while creating
```
Comment (schustmi) [inline src/zenml/integrations/mlflow/mlflow_utils.py:53]:
```suggestion
    # TODO [MEDIUM]: find a solution to avoid race-conditions while creating
```

File: src/zenml/integrations/mlflow/mlflow_utils.py [hunk 16]
```diff
@@ -11,154 +11,57 @@
...
+    IMPORTANT: this function is not race condition proof. If two or more
+    processes call it at the same time and with the same arguments, it could
+    lead to a situation where two or more MLflow runs with the same name
+    and different IDs are created.
 
     Args:
-        backend_store_uri: The mlflow backend to log to
-        experiment_name: The experiment name under which all runs will be
-                         tracked
+        experiment_name (str): the experiment name under which this runs will
```
Comment (schustmi) [inline src/zenml/integrations/mlflow/mlflow_utils.py:38]:
```suggestion
        experiment_name: the experiment name under which this runs will
```

File: src/zenml/integrations/mlflow/mlflow_utils.py [hunk 17]
```diff
@@ -11,154 +11,57 @@
...
+            be tracked. If no MLflow experiment with this name exists, one will
+            be created.
+        run_name (str): the name of the MLflow run. If a run with this name
```
Comment (schustmi) [inline src/zenml/integrations/mlflow/mlflow_utils.py:41]:
```suggestion
        run_name: the name of the MLflow run. If a run with this name
```

**Review by schustmi — CHANGES_REQUESTED** [review]
Really great changes to make the Environment much more flexible! Only a few very minor comments

File: src/zenml/environment.py (hunk 11 above)
Comment (stefannica) [reply src/zenml/environment.py]: yes, good idea

File: src/zenml/steps/step_environment.py (hunk 13 above)
Comment (stefannica) [reply src/zenml/steps/step_environment.py]: done

File: src/zenml/integrations/mlflow/mlflow_environment.py (hunk 14 above)
Comment (stefannica) [reply src/zenml/integrations/mlflow/mlflow_environment.py]: also done

**Comment by stefannica** [comment]
@schustmi thanks, I applied all your suggestions

**Review by schustmi — APPROVED** [review]
🥳
